README.txt
generic TCP server status monitoring tool.

setup the config file and when run it'll show a window
with up/down icons for the servers listed and will
periodically check their state.

HISTORY.TXT
sstat revision history
================================================================================
v1.6, 2005-12-10
    - updated contact info, added web page URL, re-released
    - updated default config file

v1.5, 2004-12-02
    - icon re-work to make the icons clearer and to add up/down arrow and
      question mark indicator, providing a second state indicator in addition
      to just color (which supports viewing by someone who can't distinguish
      red/green)
    - found and fixed a handle leak which would cause all servers to show as
      down after running sstat long enough with sufficient server downtime
    - when the main window is minimized, it now gets hidden (doesn't show on
      windows desktop toolbar).  clicking on the tray icon will restore it
    - updated a number of server IPs for the wow config files


src/sstat.h
//******************************************************************************
// sstat server monitoring tool Copyright (C) 2004 coder_1024
// (Email: coder_1024@hotmail.com)
// 
// This library is free software; you can redistribute it and/or modify it under
// the terms of the GNU Lesser General Public License as published by the Free
// Software Foundation; either version 2.1 of the License, or (at your option) any
// later version.
// 
// This library is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
// PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public License along
// with this library; if not, write to the
// 
// Free Software Foundation, Inc.,
// 59 Temple Place, Suite 330,
// Boston, MA 02111-1307 USA
//******************************************************************************

// sstat.h : main header file for the PROJECT_NAME application
//

#pragma once

#ifndef __AFXWIN_H__
	#error include 'stdafx.h' before including this file for PCH
#endif

#include "resource.h"		// main symbols


// CsstatApp:
// See sstat.cpp for the implementation of this class
//

class CsstatApp : public CWinApp
{
public:
	CsstatApp();

// Overrides
	public:
	virtual BOOL InitInstance();

// Implementation

	DECLARE_MESSAGE_MAP()
};

extern CsstatApp theApp;

src/stdafx.h
//******************************************************************************
// sstat server monitoring tool Copyright (C) 2004 coder_1024
// (Email: coder_1024@hotmail.com)
// 
// This library is free software; you can redistribute it and/or modify it under
// the terms of the GNU Lesser General Public License as published by the Free
// Software Foundation; either version 2.1 of the License, or (at your option) any
// later version.
// 
// This library is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
// PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public License along
// with this library; if not, write to the
// 
// Free Software Foundation, Inc.,
// 59 Temple Place, Suite 330,
// Boston, MA 02111-1307 USA
//******************************************************************************

// stdafx.h : include file for standard system include files,
// or project specific include files that are used frequently,
// but are changed infrequently

#pragma once

#ifndef VC_EXTRALEAN
#define VC_EXTRALEAN		// Exclude rarely-used stuff from Windows headers
#endif

// Modify the following defines if you have to target a platform prior to the ones specified below.
// Refer to MSDN for the latest info on corresponding values for different platforms.
#ifndef WINVER				// Allow use of features specific to Windows 95 and Windows NT 4 or later.
#define WINVER 0x0400		// Change this to the appropriate value to target Windows 98 and Windows 2000 or later.
#endif

#ifndef _WIN32_WINNT		// Allow use of features specific to Windows NT 4 or later.
#define _WIN32_WINNT 0x0400		// Change this to the appropriate value to target Windows 98 and Windows 2000 or later.
#endif						

#ifndef _WIN32_WINDOWS		// Allow use of features specific to Windows 98 or later.
#define _WIN32_WINDOWS 0x0410 // Change this to the appropriate value to target Windows Me or later.
#endif

#ifndef _WIN32_IE			// Allow use of features specific to IE 4.0 or later.
#define _WIN32_IE 0x0400	// Change this to the appropriate value to target IE 5.0 or later.
#endif

#define _ATL_CSTRING_EXPLICIT_CONSTRUCTORS	// some CString constructors will be explicit

// turns off MFC's hiding of some common and often safely ignored warning messages
#define _AFX_ALL_WARNINGS

#include <afxwin.h>         // MFC core and standard components
#include <afxext.h>         // MFC extensions
#include <afxdisp.h>        // MFC Automation classes

#include <afxdtctl.h>		// MFC support for Internet Explorer 4 Common Controls
#ifndef _AFX_NO_AFXCMN_SUPPORT
#include <afxcmn.h>			// MFC support for Windows Common Controls
#endif // _AFX_NO_AFXCMN_SUPPORT

#include <winsock2.h>
#include <mmsystem.h>


src/sstat.cpp
//******************************************************************************
// sstat server monitoring tool Copyright (C) 2004 coder_1024
// (Email: coder_1024@hotmail.com)
// 
// This library is free software; you can redistribute it and/or modify it under
// the terms of the GNU Lesser General Public License as published by the Free
// Software Foundation; either version 2.1 of the License, or (at your option) any
// later version.
// 
// This library is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
// PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public License along
// with this library; if not, write to the
// 
// Free Software Foundation, Inc.,
// 59 Temple Place, Suite 330,
// Boston, MA 02111-1307 USA
//******************************************************************************

// sstat.cpp : Defines the class behaviors for the application.
//

#include "stdafx.h"
#include "sstat.h"
#include "sstatDlg.h"

#ifdef _DEBUG
#define new DEBUG_NEW
#endif


// CsstatApp

BEGIN_MESSAGE_MAP(CsstatApp, CWinApp)
	ON_COMMAND(ID_HELP, CWinApp::OnHelp)
END_MESSAGE_MAP()


// CsstatApp construction

CsstatApp::CsstatApp()
{
	// TODO: add construction code here,
	// Place all significant initialization in InitInstance
}


// The one and only CsstatApp object

CsstatApp theApp;


// CsstatApp initialization

BOOL CsstatApp::InitInstance()
{
	// InitCommonControls() is required on Windows XP if an application
	// manifest specifies use of ComCtl32.dll version 6 or later to enable
	// visual styles.  Otherwise, any window creation will fail.
	InitCommonControls();

	CWinApp::InitInstance();

	AfxEnableControlContainer();

	// Standard initialization
	// If you are not using these features and wish to reduce the size
	// of your final executable, you should remove from the following
	// the specific initialization routines you do not need
	// Change the registry key under which our settings are stored
	// TODO: You should modify this string to be something appropriate
	// such as the name of your company or organization
	SetRegistryKey(_T("coder_1024"));

	CsstatDlg dlg;
	m_pMainWnd = &dlg;
	INT_PTR nResponse = dlg.DoModal();
	if (nResponse == IDOK)
	{
		// TODO: Place code here to handle when the dialog is
		//  dismissed with OK
	}
	else if (nResponse == IDCANCEL)
	{
		// TODO: Place code here to handle when the dialog is
		//  dismissed with Cancel
	}

	// Since the dialog has been closed, return FALSE so that we exit the
	//  application, rather than start the application's message pump.
	return FALSE;
}


src/resource.h
//{{NO_DEPENDENCIES}}
// Microsoft Visual C++ generated include file.
// Used by sstat.rc
//
#define IDM_ABOUTBOX                    0x0010
#define IDD_ABOUTBOX                    100
#define IDS_ABOUTBOX                    101
#define IDD_SSTAT_DIALOG                102
#define IDR_MAINFRAME                   128
#define IDB_UP                          129
#define IDB_DOWN                        130
#define IDI_ICON_UNK                    130
#define IDB_UNKNOWN1                    131
#define IDB_UNKNOWN                     131
#define IDI_ICON_DOWN                   131
#define IDI_ICON_UP                     132
#define IDB_DOWN_SEL                    132
#define IDB_UNKNOWN_SEL                 133
#define IDB_UP1                         134
#define IDB_UP_SEL                      134
#define IDC_STATUSLIST                  1000

// Next default values for new objects
// 
#ifdef APSTUDIO_INVOKED
#ifndef APSTUDIO_READONLY_SYMBOLS
#define _APS_NEXT_RESOURCE_VALUE        131
#define _APS_NEXT_COMMAND_VALUE         32771
#define _APS_NEXT_CONTROL_VALUE         1001
#define _APS_NEXT_SYMED_VALUE           101
#endif
#endif


src/stdafx.cpp
//******************************************************************************
// sstat server monitoring tool Copyright (C) 2004 coder_1024
// (Email: coder_1024@hotmail.com)
// 
// This library is free software; you can redistribute it and/or modify it under
// the terms of the GNU Lesser General Public License as published by the Free
// Software Foundation; either version 2.1 of the License, or (at your option) any
// later version.
// 
// This library is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
// PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public License along
// with this library; if not, write to the
// 
// Free Software Foundation, Inc.,
// 59 Temple Place, Suite 330,
// Boston, MA 02111-1307 USA
//******************************************************************************

// stdafx.cpp : source file that includes just the standard includes
// sstat.pch will be the pre-compiled header
// stdafx.obj will contain the pre-compiled type information

#include "stdafx.h"




src/sstatDlg.h
//******************************************************************************
// sstat server monitoring tool Copyright (C) 2004 coder_1024
// (Email: coder_1024@hotmail.com)
// 
// This library is free software; you can redistribute it and/or modify it under
// the terms of the GNU Lesser General Public License as published by the Free
// Software Foundation; either version 2.1 of the License, or (at your option) any
// later version.
// 
// This library is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
// PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public License along
// with this library; if not, write to the
// 
// Free Software Foundation, Inc.,
// 59 Temple Place, Suite 330,
// Boston, MA 02111-1307 USA
//******************************************************************************

// sstatDlg.h : header file
//

#pragma once
#include "afxcmn.h"

#include <string>
#include <list>
#include <vector>
using namespace std;

class mutex
{
public:
    mutex()
    {
        InitializeCriticalSection(&critSec);
    }

    ~mutex()
    {
        unlock();
    }

    void lock()
    {
        EnterCriticalSection(&critSec);
    }

    void unlock()
    {
        LeaveCriticalSection(&critSec);
    }

protected:
    CRITICAL_SECTION critSec;
};

class mutexguard
{
public:
    mutexguard(mutex & m)
        : m(m)
    {
        m.lock();
    }

    ~mutexguard()
    {
        m.unlock();
    }

protected:
    mutex & m;
};

typedef enum ServerStatus
{
    SS_UNKNOWN = 0,
    SS_DOWN    = 1,
    SS_UP      = 2
};

#define LEN_SERVERSTATUS 3

class ServerInfo
{
public:
    ServerInfo()
    {
        port = -1;
        status = SS_UNKNOWN;
        lastUpdate = currentUpdate = 0;
        itemIdx = -1;
        thread = NULL;
        socket = INVALID_SOCKET;
    }

    string displayName;
    string host;
    int port;

    ServerStatus status;
    long lastUpdate, currentUpdate;
    int itemIdx;

    HANDLE thread;
    SOCKET socket;
    mutex m;

};

// CsstatDlg dialog
class CsstatDlg : public CDialog
{
// Construction
public:
    CsstatDlg(CWnd* pParent = NULL);    // standard constructor

// Dialog Data
    enum { IDD = IDD_SSTAT_DIALOG };

    protected:
    virtual void DoDataExchange(CDataExchange* pDX);    // DDX/DDV support


// Implementation
protected:
    int sldx, sldy;
    list<ServerInfo*> servers;
    ServerInfo * selectedServer;
    long updatePeriod, timeout;
    long lastUpdate;
	static bool enableSounds;
	static string initialSound, downSound, upSound;
	bool enableLogging;
	string logFile;
    string configFile;
    HANDLE monitorThread;
    
    string instanceTitle;
    NOTIFYICONDATA nid;

    CBitmap m_bmpUnknown, m_bmpDown, m_bmpUp;
    CBitmap m_bmpUnknownSel, m_bmpDownSel, m_bmpUpSel;
    CImageList m_imgList;

    HICON trayIcons[LEN_SERVERSTATUS];

    bool loadConfig(CString & errText);
    vector<string> tokenizeString(char * str, char * sep);

    static DWORD WINAPI MonitorThread(LPVOID lpParm);
    static DWORD WINAPI ServerStatusThread(LPVOID lpParm);

    void LogServerStates();

    HICON m_hIcon;

    // Generated message map functions
    virtual BOOL OnInitDialog();
    afx_msg void OnSysCommand(UINT nID, LPARAM lParam);
    afx_msg void OnPaint();
    afx_msg HCURSOR OnQueryDragIcon();
    DECLARE_MESSAGE_MAP()
public:
    CListCtrl m_statusList;
    afx_msg void OnSize(UINT nType, int cx, int cy);
    afx_msg void OnDestroy();
    afx_msg void OnNMDblclkStatuslist(NMHDR *pNMHDR, LRESULT *pResult);
protected:
    virtual LRESULT WindowProc(UINT message, WPARAM wParam, LPARAM lParam);
};
